src/pycldf/orm.py

```diff
@@ -134,7 +145,11 @@ def related(self, relation):
                 '{} is list-valued, use `all_related` to retrieve related objects'.format(relation))
         fk = getattr(self.cldf, relation, None)
         if fk:
-            return self.dataset.get_object(TERMS[relation].references, fk)
+            ref = self.dataset.get_foreign_key_reference(self.component_name(), relation)
+            if ref:
+                if str(ref[1].propertyUrl) == term_uri('id'):
+                    return self.dataset.get_object(TERMS[relation].references, fk)
+                return self.dataset.get_object(TERMS[relation].references, fk, pk=True)
```

Comment:
@Anaphory  I realize that supporting non-id primary keys is a bit cumbersome. But with a bit more convenient APIs in [csvw](https://github.com/cldf/csvw/issues/56) and `pycldf` this seems manageable. Of course, it raises the bar for re-implementations, e.g. `rcldf` or similar. OTOH, stating a limitation like "primary keys must be id columns" seems simple enough to understand.

Reply:
This code looks like we assume that a foreign key always needs to point to a primary key, no matter whether that primary key is the ID column or not?
I thought I had understood that foreign keys can, at least in theory, also point to some other non-ID non-PK column. There's a call-out concerning that in the CSVW context:
https://www.w3.org/TR/2015/REC-tabular-metadata-20151217/#foreign-key-reference-between-schemas, Section 5.5, which as far as I understand we inherit indirectly, states
> Note: It is not required for the table or schema referenced from a foreignKeys property to have a similarly defined primaryKey, though frequently it will.
(but it MUST be enough to uniquely identify the corresponding row.)
So, just going by ID falling back to primary key is technically not enough, it would have to actually check by the named column.

(There are a few other strange things I noticed over there: Such as that I don't properly understand column names – which appear to be intended as within the metadata only, not leaking out – vs. column titles, or that we MUST check that foreign key references uniquely identify rows, but those are not for this PR to worry about, just for me to note down in case I forget.)

Reply:
(Also, this side note showed me once again how much more cool functionality is already in pycldf, which I probably re-implemented naïvely because I am bad at reading up on it and started diving into CLDF before it was a thing. I should figure out how to catch up.)

Reply:
Yes, that's a bit of a limitation of pycldf, which should be documented somewhere. But since CSVW doesn't have a `unique` keyword or other means to declare a column as suitable for a foreign key, we only support primary key or id here. This makes sense, since we make the objects accessible in a `dict`, and want to rely on the checks in `validate`.

Reply:
Can we please just bake that constraint directly into the CLDF specs? ‘The target of a foreign key MUST be the ID or primary key of the column’? All this weirdness that comes from the fact that you might pull data from somewhere unrelated on the web, and therefore you need to be prepared for all sorts of crazyness, shouldn't have to apply to us.

Reply:
Or is this one of the cases you already found a different use case for in CLTS?

Reply:
I think I'm ok with doing this for CLDF reference properties, because for these we know that we are not having to deal with composite keys. Imposing this constraint in the spec clearly breaks backwards compatibility, though - if only in theory, because there are no datasets in the wild, violating this constraint. The experience from CLTS is in line with such a constraint, as is the experience with the clld database schema - where all tables have an integer, surrogate primary key, and a meaningful, unique id column.

Reply:
OTOH, figuring out whether a dataset violates this constraint isn't particularly hard. So "supporting" this extra-constraint in CLDF consumers by raising "NotImplemented" if the constraint isn't met, doesn't seem like a bad solution.

Reply:
Sounds good, shall we add that NotImplementedError (NotImplemented is the execption for ‘Self Operator Other isn't implemented, try Other ROperator Self’, isn't it? I always confuse them) to the bit of code we are discussing here?

Reply:
yes.